### DISCUSSION TITLE
Keyboard emulation for macos and windows

### DISCUSSION THREAD

**Discussion by AnnikaJoo** [AvaloniaUI/Avalonia#11588] [Q&A]
Hello all, 
I am looking for a way to simulate writing on a keyboard in my application. I tried to use the Nuget package Desktop.Robot for this (https://github.com/lucassklp/Desktop.Robot). Unfortunately this only works on Windows and on macos the application crashes. I have already opened an issue on Github describing the problem under macos (https://github.com/lucassklp/Desktop.Robot/issues/36). 

With Windows.Forms the keyboard simulation works with the help of the SendKeys method (https://learn.microsoft.com/de-de/dotnet/api/system.windows.forms.sendkeys.send?view=windowsdesktop-7.0)
I would now like to implement this approach for Windows and Macos in Avalonia. 

Does anyone of you have an idea how to simulate such a keyboard input programmatically in Avalonia? I am grateful for any tip :)

**Comment by maxkatz6** [comment]
You can only send key events to the specific avalonia control/window using `target.RaiseEvent(new KeyEventArgs { Event = KeyDownEvent, Key = Key.Event } )`. There is no cross-platform way to emulate key presses OS wide (as Desktop.Robot does it for Windows).

**Comment by SimonORorke** [comment]
@maxkatz6 , The technique you suggested may solve a problem I'm working on. I want to programmatically focus a `TabItem` of a `TabControl` once the `MainWindow` has been shown.  But the `TabItem`'s `Focus `method throws a `NullRererenceException`. So emulating Tab key presses instead sounds promising.

I have not been able to get it to work though. `KeyEventArgs` does not have the `Event` property you showed. So I've tried variations of the following:
```csharp
public MainWindow() {
    InitializeComponent();
    Dispatcher.UIThread.Post(() => {
        RaiseEvent(new KeyEventArgs {
            Key = Key.Tab,
            RoutedEvent = new RoutedEvent(nameof(KeyDownEvent), RoutingStrategies.Bubble,
                typeof(KeyEventArgs), typeof(Window))
        });
    });
}
```
But that has no effect.  Neither does this:
```csharp
public MainWindow() {
    InitializeComponent();
    Dispatcher.UIThread.Post(() => {
        OnKeyDown(new KeyEventArgs {Key = Key.Tab});
    });
}
```
I would appreciate any suggestions.

**Reply by timunie** [reply]
Focus the TabItem in OnLoaded override instead of the constructor. To this time the reference shouldn't be null anymore. You can also stick with Dispatcher, but set priority to Loaded.

**Reply by SimonORorke** [reply]
Hello @timunie. Thanks for your good advice. It's surely more elegant to do this in `OnLoaded`.  I still don't have it working, but I have spotted my mistake:
```csharp
protected override void OnLoaded(RoutedEventArgs e) {
    base.OnLoaded(e);
    Debug.WriteLine(TabControl.Items[0]!.GetType().Name);
    // The item is a TabItemViewModel, not a TabItem.
    // So this cast will return null, which the ! at the end obscures.
    var firstTabItem = (TabControl.Items[0] as TabItem)!;
    // And so this compiles but throws a NullReferenceException.
    firstTabItem.Focus();
}
```
My confusion was not helped by the fact that, until I changed it just now, I had unwisely named the view model `TabItem`. It was not a problem till I needed to access `Avalonia.Controls.TabItem`!

So my question now is how to I access the `TabControl`'s `TabItems`?

**Reply by timunie** [reply]
In Avalonia.Samples we have some sort if ViewInteraction. The DialogManager may be suitable to register a View for an ItemViewModel and activate it. Or use VisualTreeExtensions to query child controls.

**Reply by SimonORorke** [reply]
Thanks ! `VisualExtensions` worked for me:
```csharp
protected override void OnLoaded(RoutedEventArgs e) {
  var firstTabItem = TabControl.FindDescendantOfType<TabItem>();
  firstTabItem!.Focus(NavigationMethod.Tab);
}
```
